src/libstd/io/mod.rs

```diff
@@ -0,0 +1,928 @@
+
+/// A trait for objects which are byte-oriented sink.
+///
+/// Writers are defined by one method, `write`. This function will attempt to
+/// write some data into the object, returning how many bytes were successfully
+/// written.
+///
+/// Another commonly overridden method is the `flush` method for writers such as
+/// buffered writers.
+///
+/// Writers are intended to be composable with one another. Many objects
+/// throughout the I/O and related libraries take and provide types which
+/// implement the `Write` trait.
+pub trait Write {
+    /// Write a buffer into this object, returning how many bytes were read.
+    ///
+    /// This function will attempt to write the entire contents of `buf`, but
+    /// the entire write may not succeed, or the write may also generate an
+    /// error. A call to `write` represents *at most one* attempt to write to
+    /// any wrapped object.
+    ///
+    /// Calls to `write` are not guaranteed to block waiting for data to be
+    /// written, and a write which would otherwise block is indicated through an
+    /// `Err` variant.
+    ///
+    /// If the return value is `Ok(n)` then it must be guaranteed that
+    /// `0 <= n <= buf.len()`. A return value of `0` typically means that the
+    /// underlying object is no longer able to accept bytes and will likely not
+    /// be able to in the future as well, or that the buffer provided is empty.
+    ///
+    /// # Errors
+    ///
+    /// Each call to `write` may generate an I/O error indicating that the
+    /// operation could not be completed. If an error is returned then no bytes
+    /// in the buffer were successfully written to this writer.
+    ///
+    /// It is **not** considered an error if the entire buffer could not be
+    /// written to this writer.
+    fn write(&mut self, buf: &[u8]) -> Result<usize>;
+
+    /// Flush this output stream, ensuring that all intermediately buffered
+    /// contents reach their destination.
+    ///
+    /// This is by default a no-op and implementers of the `Write` trait should
+    /// decide whether their stream needs to be buffered or not.
+    fn flush(&mut self) -> Result<()> { Ok(()) }
+
+    /// Attempts to write an entire buffer into this write.
+    ///
+    /// This method will continuously call `write` while there is more data to
+    /// write. This method will not return until the entire buffer has been
+    /// successfully written or an error occurs. The first error generated from
+    /// this method will be returned.
+    ///
+    /// # Errors
+    ///
+    /// This function will return the first error that `write` returns.
+    fn write_all(&mut self, mut buf: &[u8]) -> Result<()> {
+        while buf.len() > 0 {
+            match self.write(buf) {
+                Ok(0) => return Err(Error::new(ErrorKind::EndOfFile,
+                                               "failed to write whole buffer: \
+                                                eof reached", None)),
```

Comment:
Currently this is used to signify that the `write` could not complete because the "destination was full", which in some sense is a bit of an EOF condition. For example the main user of this is the `Write` impl for `&mut [u8]` which returns EOF when `self` is the empty slice.

That being said though, I can definitely see where having an `EndOfFile` variant of `ErrorKind` is somewhat confusing as it's not intended to be returned for functions like `read`.